### DISCUSSION TITLE
jotai@2.5.0 + jotai-effect@0.2.1 bug

### DISCUSSION THREAD

**Discussion by dgeibi** [pmndrs/jotai#2218] [Bug report] [answered]
found loggingEffect only triggered once

```js
import { atom, useAtom, useAtomValue } from "jotai";
import { atomEffect } from "jotai-effect";

const countAtom = atom(0);

const autoCountEffect = atomEffect((get, set) => {
  const timer = setInterval(() => {
    set(countAtom, (prev) => prev + 1);
  }, 1e3);
  return () => {
    clearInterval(timer);
  };
});

const loggingEffect = atomEffect((get) => {
  // jotai@2.4.3 log 0...1...2...
  // jotai@2.5.0 log 0 only
  console.log(get(countAtom));
});

export default function App() {
  useAtom(autoCountEffect);
  useAtom(loggingEffect);
  return useAtomValue(countAtom);
}
```

## Link to reproduction

https://codesandbox.io/p/sandbox/flamboyant-yonath-wp2rx8

**Comment by dmaskasky** [comment]
I've opened https://github.com/jotaijs/jotai-effect/issues/13 on this issue.

I reran all jotai-effect tests with jotai@2.5.0 and confirmed that all 19 tests pass. Not sure how this was missed.

The first step is to write a test that reproduces this issue. Admittedly, I'm having difficulty reproducing the issue outside the codesandbox example.

A very interesting behavior is your codesandbox example works if get hot reload to refresh.

**Comment by dmaskasky** [comment]
https://github.com/pmndrs/jotai/assets/5034873/7341e33f-b095-43b7-b11a-bbbb14d17ef3

**Comment by dmaskasky** [comment]
@dai-shi https://github.com/pmndrs/jotai/compare/v2.4.3...v2.5.0

**Reply by dai-shi** [reply]
https://github.com/pmndrs/jotai/releases/tag/v2.5.0

#2186 working
#2197 working
#2199 not working https://codesandbox.io/s/friendly-artem-vk4xc6?file=/package.json

**Reply by dai-shi** [reply]
A simpler reproduction: https://codesandbox.io/s/late-forest-fw5v2w?file=/src/App.tsx

```tsx
import { atom, useAtom } from "jotai";
import { atomEffect } from "jotai-effect";

const countAtom = atom(0);

const loggingEffect = atomEffect((get) => {
  console.log(get(countAtom));
});

export default function App() {
  useAtom(loggingEffect);
  const [count, setCount] = useAtom(countAtom);
  return (
    <>
      {count} <button onClick={() => setCount((c) => c + 1)}>+1</button>
    </>
  );
}
```

**Comment by dmaskasky** [comment]
A more simple example. The effect does not recompute when dependencies change.
https://codesandbox.io/s/relaxed-neco-t7cjyj?file=/src/App.tsx

**Comment by dmaskasky** [comment]
Added this PR that adds tests written three ways, but tests pass whereas codesandbox fails ¯\_(ツ)_/¯
https://github.com/jotaijs/jotai-effect/pull/14

**Reply by dai-shi** [reply]
Reproduced and fixed.
https://codesandbox.io/s/modest-chaplygin-3jqg2f
https://codesandbox.io/s/pedantic-chatelet-6ymlf4

**Reply by dmaskasky** [reply]
Looks like the issue was strict mode, but the code change in atomEffect looks to be nearly functionally equivalent. The only difference being that when `!ref.mounted` or when `ref.inProgress` is the result of a call to set, that undefined is returned instead of a promise.

I don't understand how this bug relates to strict mode and how this change fixes the bug.

**Reply by dmaskasky** [reply]
```ts
// atomEffect.test.ts
const effectAtom = atomEffect((get) => {
  runCount++
  console.log('[run]', count, { runCount })
  get(countAtom)
})
```

oh, we need to return the original promise when effect runs twice in. Hmm.

**Before Fix With Strict Mode**
```ts
// atomEffect.ts
if (!ref.mounted || ref.inProgress) {
  console.log('[returning early]', count)
  return ref.promise
}
```

```
  [returning early] 1
  [continue] 2
  [returning early] 3
  (run) 3 { runCount: 1 }
  (increment) 3
```

**After Fix With Strict Mode**
```ts
// atomEffect.ts
if (!ref.mounted || ref.inProgress) {
  console.log('[returning early]', count, ref.promise)
  return ref.promise
}
```

```
  [returning early] 1 undefined
  [continue] 2
  [returning early] 3 Promise { <pending>, count: 2 }
  [run] 3 { runCount: 1 }
  [continue] 4
  [run] 4 { runCount: 2 }
```

**Before Fix Without Strict Mode**
```
  [returning early] 1
  [continue] 2
  (run) 2 { runCount: 1 }
  (increment) 2
  [continue] 3
  (run) 3 { runCount: 2 }
```

**After Fix Without Strict Mode**
```
  [returning early] 1 undefined
  [continue] 2
  [run] 2 { runCount: 1 }
  [continue] 3
  [run] 3 { runCount: 2 }
```

**Comment by dmaskasky** [comment, marked as answer]
The root cause is strict mode in dev mode causes the effect to run multiple times which wasn't supported by the code.
This issue has been resolved with https://github.com/jotaijs/jotai-effect/pull/14.
I also duplicated the full test suite to run in strict mode. Now every test has non-strict and strict mode variants.

**Comment by M-Nasser-M** [comment]
this issue still exists for me link to reproduction 
https://codesandbox.io/s/bitter-water-d6zl87?file=%2Fpackage.json

**Reply by dmaskasky** [reply]
I just pushed `v0.2.2` which includes the fix.

Your codesandbox is now working.
https://codesandbox.io/s/compassionate-nash-mml269

**Reply by M-Nasser-M** [reply]
yeah i just tried it on my local project and everything is good thanks alot ♥
